include/sys/dsl_dataset.h

```diff
@@ -165,6 +166,7 @@ typedef struct dsl_dataset {
 	uint64_t ds_object;
 	uint64_t ds_fsid_guid;
 	boolean_t ds_is_snapshot;
+	struct dsl_key_mapping *ds_key_mapping;
```

Comment:
Is there any mechanism for setting this back to NULL when we release it down to 0 holds?

Reply:
Unfortunately not, although now with the new `key_mapping_rele()` function I think I can do that... I'll add it in now.

Reply:
Actually I don't think it will work after all without introducing a ton of locking / unlocking.... Do you have any ideas for how to do that?